ably/ablytest/recorders.go

```diff
@@ -125,168 +123,78 @@ func (rec *RoundTripRecorder) roundTrip(req *http.Request) (*http.Response, erro
-// result in some of them being dropped.
-func NewStateRecorder(buffer int) *StateRecorder {
-	return newStateRecorder(buffer, ably.StateChan|ably.StateConn)
+	states []ably.ConnectionState
 }
 
-// NewStateChanRecorder gives new recorder which records channel-related
-// state transitions only.
-func NewStateChanRecorder(buffer int) *StateRecorder {
-	return newStateRecorder(buffer, ably.StateChan)
+func (cs *ConnStatesRecorder) append(state ably.ConnectionState) {
+	cs.mtx.Lock()
+	defer cs.mtx.Unlock()
+	cs.states = append(cs.states, state)
 }
 
-// NewStateConnRecorder gives new recorder which records connection-related
-// state transitions only.
-func NewStateConnRecorder(buffer int) *StateRecorder {
-	return newStateRecorder(buffer, ably.StateConn)
-}
-
-func newStateRecorder(buffer int, typ ably.StateType) *StateRecorder {
-	if buffer < 0 {
-		buffer = 0
-	}
-	rec := &StateRecorder{
-		ch:     make(chan ably.State, buffer),
-		done:   make(chan struct{}),
-		states: make([]ably.State, 0, buffer),
-		typ:    typ,
-	}
-	rec.wg.Add(1)
-	go rec.processIncomingStates()
-	return rec
+func (cs *ConnStatesRecorder) Listen(r *ably.Realtime) (off func()) {
+	cs.append(r.Connection.State())
+	return r.Connection.OnAll(func(c ably.ConnectionStateChange) {
+		cs.append(c.Current)
+	})
 }
 
-func (rec *StateRecorder) processIncomingStates() {
-	defer rec.wg.Done()
-	for {
-		select {
-		case state, ok := <-rec.ch:
-			if !ok {
-				return
-			}
-			if state.Type != 0 && state.Type&rec.typ == 0 {
-				continue
-			}
-			rec.add(state)
-		case <-rec.done:
-			return
-		}
-	}
+func (cs *ConnStatesRecorder) States() []ably.ConnectionState {
+	cs.mtx.Lock()
+	defer cs.mtx.Unlock()
+	return cs.states
 }
 
-// Add appends state to the list of recorded ones, used to ensure ordering
-// of the states by injecting values at certain points of the test.
-func (rec *StateRecorder) Add(state ably.StateEnum) {
-	rec.ch <- ably.State{State: state}
+type ChanStatesRecorder struct {
+	mtx    sync.Mutex
+	states []ably.ChannelState
 }
 
-func (rec *StateRecorder) add(state ably.State) {
-	rec.mtx.Lock()
-	rec.states = append(rec.states, state)
-	rec.mtx.Unlock()
+func (cs *ChanStatesRecorder) append(state ably.ChannelState) {
+	cs.mtx.Lock()
+	defer cs.mtx.Unlock()
+	cs.states = append(cs.states, state)
 }
 
-func (rec *StateRecorder) Channel() chan<- ably.State {
-	return rec.ch
+func (cs *ChanStatesRecorder) Listen(channel *ably.RealtimeChannel) (off func()) {
+	cs.append(channel.State())
+	return channel.OnAll(func(c ably.ChannelStateChange) {
+		cs.append(c.Current)
+	})
 }
 
-// Stop stops separate recording goroutine and waits until it terminates.
-func (rec *StateRecorder) Stop() {
-	close(rec.done)
-	rec.wg.Wait()
-	// Drain listener channel.
-	for {
-		select {
-		case <-rec.ch:
-		default:
-			return
-		}
-	}
+func (cs *ChanStatesRecorder) States() []ably.ChannelState {
+	cs.mtx.Lock()
+	defer cs.mtx.Unlock()
+	return cs.states
```

Comment:
This isn't really necessary, because we're only appending to and reading from the slice concurrently.

An append will either write on the slice's underlying array past the previous last element (so the elements that can be read through the old slice won't be modified), or create a new underlying array and copy the old one (so the old slice will just be read for copying, not modified; and concurrent read is safe).

We only need a lock on the slice variable we're reassigning (in this case, the struct field), not on the slice itself.